DependencyResolution/PostSharp.Samples.DependencyResolution.Dynamic/Program.cs

```diff
@@ -0,0 +1,98 @@
+﻿using System;
+using System.ComponentModel.Composition;
+using System.ComponentModel.Composition.Hosting;
+using System.ComponentModel.Composition.Primitives;
+using PostSharp.Aspects;
+using PostSharp.Extensibility;
+using PostSharp.Serialization;
+
+namespace PostSharp.Samples.DependencyResolution.Dynamic
+{
+    public interface ILogger
+    {
+        void Log(string message);
+    }
+
+    public static class AspectServiceLocator
+    {
+        private static CompositionContainer container;
+        private static bool isCacheable;
+
+        public static void Initialize(ComposablePartCatalog catalog, bool isCacheable)
+        {
+            if (AspectServiceLocator.isCacheable && container != null)
+                throw new InvalidOperationException();
+
+            container = new CompositionContainer(catalog);
+            AspectServiceLocator.isCacheable = isCacheable;
+        }
+
+        public static Func<T> GetService<T>() where T : class
+        {
+            if (isCacheable)
+            {
+                return () => new Lazy<T>(GetServiceImpl<T>).Value;
```

Comment:
Nice :) It may be worth mentioning in a comment that you're using a compiler trick to make things cacheable. It didn't appear to me at first reading.